### PR TITLE
test: account compression

### PR DESCRIPTION
**Pull request by ananas-block**
changes:
- add associated queue and tree check to update address Merkle tree
- init address merkle tree: add standard config checks at init
- test: failing tests update address Merkle tree
- test: append
- assert_rpc_error returns Result<(), RpcError>
- add NewElementNextIndexMismatch error
- add checks for default value initialization in Address Merkle tree
- add check in insert into queue to throw an error on empty input values
- merkle_tree_test
  - moved all helper functions to the bottom -> diff is awful to look at
- add failing test for NewElementNextIndexMismatch in indexedMerkle tree
- fix address Merkle tree meta data assert (make index variable)

### CODE REVIEW

File: programs/account-compression/src/processor/initialize_address_merkle_tree.rs [hunk 1]
```diff
@@ -19,17 +18,12 @@ pub fn process_initialize_address_merkle_tree(
...
 
-    let rollover_fee = match rollover_threshold {
-        Some(rollover_threshold) => {
-            compute_rollover_fee(rollover_threshold, height, rent).map_err(ProgramError::from)?
-        }
-        None => 0,
-    };
-
+    // The address Merkle tree is never directly called by the user.
+    // The whole rollover fees are collected by the address queue.
```
Comment (SwenSchaeferjohann) [inline programs/account-compression/src/processor/initialize_address_merkle_tree.rs:25]:
```suggestion
    // All rollover fees are collected by the address queue.
```

**Comment by SwenSchaeferjohann** [comment]
just some nits from my side

File: merkle-tree/indexed/src/lib.rs [hunk 2]
```diff
@@ -220,19 +220,19 @@ where
                 return Err(IndexedMerkleTreeError::NewElementGreaterOrEqualToNextElement);
             }
         }
-
+        if new_element.next_index != low_element.next_index {
```
[the remark this answers is not in the document]
Comment (ananas-block) [reply merkle-tree/indexed/src/lib.rs:223]: A better place for a comment explaining how the update works could be above the update method.
Such explanation will also naturally point out why such checks are necessary.

[the remark this answers is not in the document]
Comment (ananas-block) [reply merkle-tree/indexed/src/lib.rs:223]: I will add that comment when reviewing the indexed Merkle tree crate

File: programs/account-compression/src/utils/transfer_lamports.rs [hunk 3]
```diff
@@ -0,0 +1,26 @@
+use anchor_lang::prelude::*;
+
+pub fn transfer_lamports<'info>(
+    from: &AccountInfo<'info>,
+    to: &AccountInfo<'info>,
+    lamports: u64,
+) -> Result<()> {
+    let compressed_sol_pda_lamports = from.as_ref().lamports();
```
[the remark this answers is not in the document]
Comment (ananas-block) [reply programs/account-compression/src/utils/transfer_lamports.rs:8]: Doesn't work
```
error[E0308]: mismatched types
  --> programs/account-compression/src/utils/transfer_lamports.rs:10:40
   |
10 |     *from.try_borrow_mut_lamports()? = compressed_sol_pda_lamports.checked_sub(lamports).unwrap();
   |     --------------------------------   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&mut u64`, found `u64`
   |     |
   |     expected due to the type of this binding
   |
help: consider dereferencing here to assign to the mutably borrowed value
   |
10 |     **from.try_borrow_mut_lamports()? = compressed_sol_pda_lamports.checked_sub(lamports).unwrap();
   |     +
```

File: programs/account-compression/src/utils/transfer_lamports.rs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+    **from.as_ref().try_borrow_mut_lamports()? =
```
[the remark this answers is not in the document]
Comment (ananas-block) [reply programs/account-compression/src/utils/transfer_lamports.rs:10]: ### Doesn't work
```
error[E0308]: mismatched types
  --> programs/account-compression/src/utils/transfer_lamports.rs:10:40
   |
10 |     *from.try_borrow_mut_lamports()? = compressed_sol_pda_lamports.checked_sub(lamports).unwrap();
   |     --------------------------------   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&mut u64`, found `u64`
   |     |
   |     expected due to the type of this binding
   |
help: consider dereferencing here to assign to the mutably borrowed value
   |
10 |     **from.try_borrow_mut_lamports()? = compressed_sol_pda_lamports.checked_sub(lamports).unwrap();
   |     +
```

File: programs/account-compression/src/utils/transfer_lamports.rs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        compressed_sol_pda_lamports.checked_sub(lamports).unwrap();
+    let recipient_lamports = to.as_ref().lamports();
+    **to.as_ref().try_borrow_mut_lamports()? = recipient_lamports.checked_add(lamports).unwrap();
```
[the remark this answers is not in the document]
Comment (ananas-block) [reply programs/account-compression/src/utils/transfer_lamports.rs:13]: But I can remove the unwraps

File: programs/account-compression/src/utils/transfer_lamports.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (vadorovsky) [reply programs/account-compression/src/utils/transfer_lamports.rs:11]: Fair

**Review by vadorovsky — APPROVED** [review]

**Pull request merged**